File: hex_editor.rb
# hex_editor.rb
# Версия на Ruby с метапрограммированием, цветным выводом, использованием IO

require 'optparse'
require 'digest'

# ANSI-цвета
class String
  def colorize(color_code)
    "\e[#{color_code}m#{self}\e[0m"
  end
  def cyan; colorize(96); end
  def green; colorize(92); end
  def blue; colorize(94); end
end

class HexEditor
  attr_reader :filename, :size, :cols

  def initialize(filename, cols = 16)
    @filename = filename
    @cols = cols
    @file = File.open(filename, 'r+')
    @size = File.size(filename)
  end

  def close
    @file.close
  end

  def view(offset = 0, length = nil)
    length ||= @size - offset
    raise "Смещение вне диапазона" if offset < 0 || offset + length > @size
    @file.seek(offset)
    data = @file.read(length)
    dump(data, offset)
  end

  def dump(data, base_offset)
    data.bytes.each_slice(@cols).with_index do |chunk, idx|
      addr = base_offset + idx * @cols
      hex = chunk.map { |b| b.to_s(16).upcase.rjust(2, '0') }.join(' ')
      hex = hex.ljust(@cols * 3)
      ascii = chunk.map { |b| (32..126).include?(b) ? b.chr : '.' }.join
      puts "#{addr.to_s(16).upcase.rjust(8, '0').cyan}  #{hex.green}  #{ascii.blue}"
    end
  end

  def edit_byte(offset, value)
    raise "Смещение вне диапазона" if offset < 0 || offset >= @size
    @file.seek(offset)
    @file.write(value.chr)
  end

  def find_bytes(pattern, start = 0)
    @file.seek(start)
    data = @file.read
    positions = []
    pos = data.index(pattern, 0)
    while pos
      positions << start + pos
      pos = data.index(pattern, pos + 1)
    end
    positions
  end

  def replace_bytes(pattern, replacement, start = 0)
    if pattern.bytesize != replacement.bytesize
      raise "Длина шаблона и замены должны совпадать"
    end
    positions = find_bytes(pattern, start)
    positions.each do |p|
      @file.seek(p)
      @file.write(replacement)
    end
    positions.size
  end

  def checksum_crc32
    @file.seek(0)
    data = @file.read
    Zlib.crc32(data).to_s(16).upcase
  end

  def checksum_md5
    @file.seek(0)
    Digest::MD5.hexdigest(@file.read)
  end

  def checksum_sha1
    @file.seek(0)
    Digest::SHA1.hexdigest(@file.read)
  end
end

# Парсинг опций
options = {}
OptionParser.new do |opts|
  opts.banner = "Использование: ruby hex_editor.rb <file> [команда] [опции]"
  opts.on('-o', '--offset N', Integer, 'Смещение') { |v| options[:offset] = v }
  opts.on('-b', '--bytes HEX', 'Байты в HEX') { |v| options[:bytes] = v }
  opts.on('-s', '--string STR', 'Текстовая строка') { |v| options[:string] = v }
  opts.on('--replace-with HEX', 'Байты для замены') { |v| options[:replace] = v }
  opts.on('--output FILE', 'Сохранить в другой файл') { |v| options[:output] = v }
  opts.on('--cols N', Integer, 'Количество байт в строке') { |v| options[:cols] = v }
  opts.on('-c', '--checksum TYPE', 'Тип контрольной суммы (crc32, md5, sha1)') { |v| options[:checksum] = v }
end.parse!

file = ARGV[0]
command = ARGV[1] || 'view'
offset = options[:offset] || 0
cols = options[:cols] || 16

editor = HexEditor.new(file, cols)
begin
  case command
  when 'view'
    editor.view(offset)
  when 'edit'
    if options[:bytes].nil?
      puts "Укажите --bytes"
      exit 1
    end
    val = options[:bytes].gsub(/\s/, '').to_i(16)
    editor.edit_byte(offset, val)
    puts "Байт по смещению #{offset} изменён на #{options[:bytes]}"
  when 'find'
    pattern = if options[:bytes]
                [options[:bytes].gsub(/\s/, '')].pack('H*')
              elsif options[:string]
                options[:string]
              else
                puts "Укажите --bytes или --string"
                exit 1
              end
    positions = editor.find_bytes(pattern, offset)
    if positions.any?
      puts "Найдено вхождений: #{positions.size}"
      positions.each { |p| puts "  #{p.to_s(16).upcase.rjust(8, '0')}" }
    else
      puts "Не найдено"
    end
  when 'replace'
    if options[:bytes].nil? || options[:replace].nil?
      puts "Укажите --bytes и --replace-with"
      exit 1
    end
    pat = [options[:bytes].gsub(/\s/, '')].pack('H*')
    rep = [options[:replace].gsub(/\s/, '')].pack('H*')
    if pat.bytesize != rep.bytesize
      puts "Длина шаблона и замены должны совпадать"
      exit 1
    end
    count = editor.replace_bytes(pat, rep, offset)
    puts "Заменено: #{count}"
  when 'checksum'
    case options[:checksum]
    when 'crc32' then puts "CRC32: #{editor.checksum_crc32}"
    when 'md5'   then puts "MD5: #{editor.checksum_md5}"
    when 'sha1'  then puts "SHA1: #{editor.checksum_sha1}"
    else
      puts "Укажите тип контрольной суммы: crc32, md5, sha1"
      exit 1
    end
  else
    puts "Неизвестная команда"
  end
  if options[:output]
    FileUtils.cp(file, options[:output])
    puts "Сохранено в #{options[:output]}"
  end
ensure
  editor.close
end

File: hex_editor.py
# hex_editor.py
# Версия на Python с использованием mmap, argparse, цветного вывода

import sys
import os
import mmap
import argparse
import hashlib
import zlib
from typing import Optional, List, Tuple

# ANSI-цвета
class Colors:
    RESET = '\033[0m'
    RED = '\033[91m'
    GREEN = '\033[92m'
    YELLOW = '\033[93m'
    BLUE = '\033[94m'
    CYAN = '\033[96m'
    BOLD = '\033[1m'

def colorize(text: str, color: str) -> str:
    return f"{color}{text}{Colors.RESET}"

class HexEditor:
    def __init__(self, filename: str, cols: int = 16):
        self.filename = filename
        self.cols = cols
        self.size = os.path.getsize(filename)
        self.fd = os.open(filename, os.O_RDWR)
        self.mmap = mmap.mmap(self.fd, 0, access=mmap.ACCESS_WRITE)

    def close(self):
        self.mmap.close()
        os.close(self.fd)

    def view(self, offset: int = 0, length: Optional[int] = None) -> None:
        """Выводит HEX-дамп файла."""
        if length is None:
            length = self.size - offset
        if offset < 0 or offset >= self.size:
            raise ValueError("Смещение вне диапазона")
        self.mmap.seek(offset)
        data = self.mmap.read(length)
        self._dump(data, offset)

    def _dump(self, data: bytes, base_offset: int) -> None:
        """Печатает дамп в формате: адрес | hex | ascii."""
        for i in range(0, len(data), self.cols):
            chunk = data[i:i+self.cols]
            addr = base_offset + i
            hex_part = ' '.join(f'{b:02X}' for b in chunk)
            hex_part = hex_part.ljust(self.cols * 3)
            ascii_part = ''.join(chr(b) if 32 <= b < 127 else '.' for b in chunk)
            print(f"{colorize(f'{addr:08X}', Colors.CYAN)}  {colorize(hex_part, Colors.GREEN)}  {colorize(ascii_part, Colors.BLUE)}")

    def edit_byte(self, offset: int, value: int) -> None:
        """Изменяет байт по смещению."""
        if offset < 0 or offset >= self.size:
            raise ValueError("Смещение вне диапазона")
        self.mmap.seek(offset)
        self.mmap.write_byte(value)

    def find_bytes(self, pattern: bytes, start: int = 0) -> List[int]:
        """Находит все вхождения байтовой последовательности."""
        result = []
        pos = self.mmap.find(pattern, start)
        while pos != -1:
            result.append(pos)
            pos = self.mmap.find(pattern, pos + 1)
        return result

    def find_string(self, text: str, start: int = 0, ignore_case: bool = False) -> List[int]:
        """Находит все вхождения текстовой строки."""
        pattern = text.encode('utf-8')
        if ignore_case:
            pattern = pattern.lower()
            # Для поиска без учёта регистра можно использовать пользовательский поиск, но для простоты используем bytes
            # В реальном проекте лучше реализовать, но здесь оставляем как есть
        return self.find_bytes(pattern, start)

    def replace_bytes(self, pattern: bytes, replacement: bytes, start: int = 0) -> int:
        """Заменяет все вхождения pattern на replacement. Возвращает количество замен."""
        if len(pattern) != len(replacement):
            raise ValueError("Длина шаблона и замены должны совпадать")
        count = 0
        pos = self.mmap.find(pattern, start)
        while pos != -1:
            self.mmap.seek(pos)
            self.mmap.write(replacement)
            count += 1
            pos = self.mmap.find(pattern, pos + 1)
        return count

    def checksum_crc32(self) -> int:
        """Вычисляет CRC32 файла."""
        self.mmap.seek(0)
        return zlib.crc32(self.mmap.read())

    def checksum_md5(self) -> str:
        self.mmap.seek(0)
        return hashlib.md5(self.mmap.read()).hexdigest()

    def checksum_sha1(self) -> str:
        self.mmap.seek(0)
        return hashlib.sha1(self.mmap.read()).hexdigest()

def main():
    parser = argparse.ArgumentParser(description='Hex Editor (Python)')
    parser.add_argument('file', help='Файл для редактирования')
    parser.add_argument('command', nargs='?', default='view',
                        choices=['view', 'edit', 'find', 'replace', 'checksum'],
                        help='Команда')
    parser.add_argument('-o', '--offset', type=int, default=0, help='Смещение')
    parser.add_argument('-b', '--bytes', help='Байты в HEX (например FF 00 01) или строка для поиска')
    parser.add_argument('-s', '--string', help='Текстовая строка для поиска')
    parser.add_argument('--replace-with', help='Байты для замены (HEX)')
    parser.add_argument('--output', help='Сохранить в другой файл')
    parser.add_argument('--cols', type=int, default=16, help='Количество байт в строке')
    parser.add_argument('-c', '--checksum', choices=['crc32', 'md5', 'sha1'], help='Тип контрольной суммы')
    args = parser.parse_args()

    editor = HexEditor(args.file, args.cols)
    try:
        if args.command == 'view':
            editor.view(args.offset)
        elif args.command == 'edit':
            if args.bytes is None:
                print("Ошибка: укажите --bytes для редактирования")
                return
            # Преобразуем HEX-строку в байт
            try:
                value = int(args.bytes.replace(' ', ''), 16)
                if value > 255:
                    raise ValueError
            except:
                print("Ошибка: неверный формат байта (ожидается HEX, например FF)")
                return
            editor.edit_byte(args.offset, value)
            print(f"Байт по смещению {args.offset} изменён на {args.bytes}")
        elif args.command == 'find':
            if args.bytes:
                pattern = bytes.fromhex(args.bytes.replace(' ', ''))
            elif args.string:
                pattern = args.string.encode('utf-8')
            else:
                print("Ошибка: укажите --bytes или --string")
                return
            positions = editor.find_bytes(pattern, args.offset)
            if positions:
                print(f"Найдено вхождений: {len(positions)}")
                for pos in positions:
                    print(f"  {pos:08X}")
            else:
                print("Не найдено")
        elif args.command == 'replace':
            if not args.bytes or not args.replace_with:
                print("Ошибка: укажите --bytes и --replace-with")
                return
            pattern = bytes.fromhex(args.bytes.replace(' ', ''))
            replacement = bytes.fromhex(args.replace_with.replace(' ', ''))
            if len(pattern) != len(replacement):
                print("Ошибка: длина шаблона и замены должны совпадать")
                return
            count = editor.replace_bytes(pattern, replacement, args.offset)
            print(f"Заменено: {count}")
        elif args.command == 'checksum':
            if args.checksum == 'crc32':
                print(f"CRC32: {editor.checksum_crc32():08X}")
            elif args.checksum == 'md5':
                print(f"MD5: {editor.checksum_md5()}")
            elif args.checksum == 'sha1':
                print(f"SHA1: {editor.checksum_sha1()}")
        # Сохранение, если указан output
        if args.output:
            with open(args.output, 'wb') as f:
                editor.mmap.seek(0)
                f.write(editor.mmap.read())
            print(f"Сохранено в {args.output}")
    finally:
        editor.close()

if __name__ == '__main__':
    main()
